File: webforj-webswing-integration/src/main/java/com/webforj/webswing/WebswingConnector.java
package com.webforj.webswing;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

import com.webforj.annotation.Attribute;
import com.webforj.annotation.JavaScript;
import com.webforj.component.element.ElementComposite;
import com.webforj.component.element.PropertyDescriptor;
import com.webforj.component.element.annotation.NodeName;
import com.webforj.concern.HasClassName;
import com.webforj.concern.HasSize;
import com.webforj.concern.HasStyle;
import com.webforj.dispatcher.EventListener;
import com.webforj.dispatcher.ListenerRegistration;
import com.webforj.webswing.event.WebswingConnectorInitializedEvent;
import com.webforj.webswing.event.WebswingConnectorStartedEvent;
import com.webforj.webswing.event.WebswingConnectorActionEvent;

@JavaScript(value = "ws://webswing-connector.js",
    attributes = {@Attribute(name = "type", value = "module")})
@NodeName("webswing-connector")
public class WebswingConnector extends ElementComposite implements HasSize<WebswingConnector>,
    HasStyle<WebswingConnector>, HasClassName<WebswingConnector> {

  private PropertyDescriptor<String> urlProp = PropertyDescriptor.property("url", "");
  private PropertyDescriptor<WebswingConnectorOptions> optionsProp =
      PropertyDescriptor.property("options", null);

  /**
   * Creates a new instance of the webswing connector.
   *
   * @param url the url of the webswing server
   * @param autoStart whether to start the webswing server automatically
   */
  public WebswingConnector(String url, boolean autoStart) {
    super();
    setUrl(url);
    setOptions(new WebswingConnectorOptions().setAutoStart(autoStart));
  }

  /**
   * Creates a new instance of the webswing connector.
   *
   * @param url the url of the webswing server
   */
  public WebswingConnector(String url) {
    this(url, true);
  }

  /**
   * Sets the url of the webswing server.
   *
   * @param url the url of the webswing server
   * @return the component itself
   */
  public WebswingConnector setUrl(String url) {
    set(urlProp, url);
    return this;
  }

  /**
   * Gets the url of the webswing server.
   *
   * @return the url of the webswing server
   */
  public String getUrl() {
    return get(urlProp);
  }

  /**
   * Sets the options of the webswing connector.
   *
   * @param options the options of the webswing connector
   * @return the component itself
   */
  public WebswingConnector setOptions(WebswingConnectorOptions options) {
    set(optionsProp, options);
    return this;
  }

  /**
   * Gets the options of the webswing connector.
   *
   * @return the options of the webswing connector
   */
  public WebswingConnectorOptions getOptions() {
    return get(optionsProp);
  }

  /**
   * This will initiate the connection to Webswing server and start the Swing application.
   *
   * <p>
   * If the autoStart is set to false or not defined in config object, the start method has to be
   * called manually, otherwise the Webswing will call start method automatically.
   * </p>
   *
   * @return the component itself
   */
  public WebswingConnector start() {
    getElement().callJsFunctionVoidAsync("start");
    return this;
  }

  /**
   * Perform an action that triggers server-side listener.
   *
   * @param actionName the name of the action
   * @param data the data to send to the server
   * @param binaryData the binary data to send to the server
   *
   * @return the component itself
   */
  public WebswingConnector performAction(String actionName, String data, String binaryData) {
    String encodedBinaryData = binaryData;
    if (binaryData != null) {
      encodedBinaryData =
          new String(Base64.getEncoder().encode(binaryData.getBytes(StandardCharsets.UTF_8)),
              StandardCharsets.UTF_8);
    }

    getElement().callJsFunctionVoidAsync("performAction", actionName, data, encodedBinaryData);
    return this;
  }

  /**
   * Perform an action that triggers server-side listener.
   *
   * @param actionName the name of the action
   * @param data the data to send to the server
   *
   * @return the component itself
   */
  public WebswingConnector performAction(String actionName, String data) {
    return performAction(actionName, data, "");
  }

  /**
   * Perform an action that triggers server-side listener.
   *
   * @param actionName the name of the action
   * @return the component itself
   */
  public WebswingConnector performAction(String actionName) {
    return performAction(actionName, "");
  }

  /**
   * Adds a listener for the {@link WebswingConnectorInitializedEvent} event.
   *
   * @param listener the listener
   * @return A registration object for removing the event listener
   */
  public ListenerRegistration<WebswingConnectorInitializedEvent> addInitializedListener(
      EventListener<WebswingConnectorInitializedEvent> listener) {
    return addEventListener(WebswingConnectorInitializedEvent.class, listener);
  }

  /**
   * Alias for {@link #addInitializedListener(EventListener)}.
   *
   * @param listener the listener
   * @return A registration object for removing the event listener
   */
  public ListenerRegistration<WebswingConnectorInitializedEvent> onInitialized(
      EventListener<WebswingConnectorInitializedEvent> listener) {
    return addEventListener(WebswingConnectorInitializedEvent.class, listener);
  }

  /**
   * Adds a listener for the {@link WebswingConnectorStartedEvent} event.
   *
   * @param listener the listener
   * @return A registration object for removing the event listener
   */
  public ListenerRegistration<WebswingConnectorStartedEvent> addStartedListener(
      EventListener<WebswingConnectorStartedEvent> listener) {
    return addEventListener(WebswingConnectorStartedEvent.class, listener);
  }

  /**
   * Alias for {@link #addStartedListener(EventListener)}.
   *
   * @param listener the listener
   * @return A registration object for removing the event listener
   */
  public ListenerRegistration<WebswingConnectorStartedEvent> onStarted(
      EventListener<WebswingConnectorStartedEvent> listener) {
    return addEventListener(WebswingConnectorStartedEvent.class, listener);
  }

  /**
   * Adds a listener for the {@link WebswingConnectorActionEvent} event.
   *
   * @param listener the listener
   * @return A registration object for removing the event listener
   */
  public ListenerRegistration<WebswingConnectorActionEvent> addActionListener(
      EventListener<WebswingConnectorActionEvent> listener) {
    return addEventListener(WebswingConnectorActionEvent.class, listener);
  }

  /**
   * Alias for {@link #addActionListener(EventListener)}.
   *
   * @param listener the listener
   * @return A registration object for removing the event listener
   */
  public ListenerRegistration<WebswingConnectorActionEvent> onAction(
      EventListener<WebswingConnectorActionEvent> listener) {
    return addEventListener(WebswingConnectorActionEvent.class, listener);
  }
}

File: webforj-webswing-integration/frontend/src/webswing-connector.ts
import { LitElement, html } from 'lit'
import { customElement, property } from 'lit/decorators.js'
import { getWebswingApi } from 'webswing-api';
import { IWebswingInstance, WebswingOptions } from 'webswing-api/build/types';

/**
 * A custom element that wraps a Webswing instance.
 *
 * @attr {String} url - The URL of the Webswing server.
 * @attr {Object} options - The options for initializing the Webswing instance.
 *
 * @fires webswing-initialized - Fired when the Webswing instance is initialized.
 * @fires webswing-started - Fired when the Webswing instance is started.
 * @fires webswing-action - Fired when an action is performed on the Webswing instance.
 */
@customElement("webswing-connector")
export class WebswingElement extends LitElement {
  private instance: IWebswingInstance | null = null;
  private started: boolean = false;
  private initId: number = 0;

  /**
   * The URL of the Webswing server.
   */
  @property({ type: String })
  url: string | null = null;

  /**
   * The options for initializing the Webswing instance.
   */
  @property({ type: Object })
  options: WebswingOptions | null = null;

  /**
   * {@inheritDoc}
   */
  connectedCallback() {
    super.connectedCallback();
    this.style.display = 'block';
  }

  /**
   * {@inheritDoc}
   */
  disconnectedCallback() {
    super.disconnectedCallback();
    if (this.instance != null) {
      this.instance.disconnect();
    }
    this.instance = null;
    this.started = false;
  }

  /**
   * This will initiate the connection to Webswing server and start the Swing application.
   * If the autoStart is set to false or not defined in config object, the start function has
   * to be called manually, otherwise the Webswing will call start function automatically.
   */
  public start() {
    if (this.started) {
      throw new Error("Instance already started!");
    }

    if (this.instance == null) {
      throw new Error("No instance initialized!");
    }

    this.started = true;
    this.instance.start();
  }

  /**
   * Perform an action that triggers server-side listener.
   *
   * @param actionName The name of the action to perform.
   * @param data The data to send with the action.
   * @param binaryDataBase64 The binary data to send with the action, encoded in Base64.
   */
  public performAction(actionName: string, data: string | undefined, binaryDataBase64: string | null) {
    const binaryData = this.decode(binaryDataBase64) || undefined;
    this.instance?.performAction({ actionName, data, binaryData });
  }

  /**
   * {@inheritDoc}
   */
  protected firstUpdated() {
    this.initialize();
  }

  /**
   * {@inheritDoc}
   */
  protected createRenderRoot() {
    return this;
  }

  /**
   * {@inheritDoc}
   */
  protected render() {
    return html`<div class="webswing-root"></div>`;
  }

  private initialize() {
    const url = this.url!;
    const opts = this.options;

    const thisInitId = ++this.initId;
    getWebswingApi(url).then((api) => {
      if (this.initId != thisInitId) {
        return;
      }

      let startupOptions = {
        connectionUrl: url,
        autoStart: false,
        onStart: () => {
          this.dispatchEvent(new Event('webswing-started', { bubbles: true, composed: true }));
        }
      };

      if (opts != null) {
        startupOptions = { ...startupOptions, ...opts };
      }

      this.instance = api.bootstrap(this.querySelector(".webswing-root")!, startupOptions, (injector) => {
        injector.services.base!.handleActionEvent = (actionName: string, data: string | null, binaryData: Uint8Array | null) => {
          const binaryDataBase64 = this.encode(binaryData);
          this.dispatchEvent(new CustomEvent('webswing-action', {
            detail: {
              actionName,
              data,
              binaryDataBase64
            },
            bubbles: true,
            composed: true
          }));
        }

        requestAnimationFrame(() => {
          this.dispatchEvent(new Event('webswing-initialized', { bubbles: true, composed: true }));
        });
      });
    })
      .catch((e) => {
        throw new Error(`Failed to initialize Webswing: ${e.message}`);
      });
  }

  private encode(binaryData: Uint8Array | null) {
    if (binaryData == null) {
      return null;
    }

    const decoder = new TextDecoder('utf8');
    return btoa(decoder.decode(binaryData));
  }

  private decode(binaryDataBase64: string | null) {
    if (binaryDataBase64 == null) {
      return null;
    }

    return new Uint8Array(atob(binaryDataBase64).split("").map(function (c) {
      return c.charCodeAt(0);
    }));
  }
}
